TASK: Backport the end to end tests from 8.4

markusguenther opened the pull request:
Related: https://github.com/neos/neos-ui/pull/3850
Related: https://github.com/neos/neos-ui/pull/3851

In the 8.4 branch, we worked on improving the acceptance test setup. Since we are supporting both Neos 8.3 and 8.4 simultaneously, the new test setup includes updated versions of TestCafe and SauceLabs.

This PR just backports the functionality.

The e2e.sh got some new parameters to be able to run local tests and saucelabs. Additionally, we introduce a help parameter to get more details on the usage.

```e2e.sh --help```
<img width="695" alt="Screenshot 2024-09-24 at 20 58 25" src="https://github.com/user-attachments/assets/53f3c538-8a4e-44b7-973e-70a1fb47c75d">

For the saucelabs usage we need the environment variables `SAUCE_USERNAME` and `SAUCE_ACCESS_KEY`
```e2e.sh --saucelabs```

Local usage with e.g. the chrome browser works like that
```e2e.sh --browser chrome```

skurfuerst committed:
Merge pull request #3850 from neos/84-fix-e2e-tests

8.4: fix e2e tests
# Conflicts:
#	packages/neos-ui-redux-store/src/CR/Nodes/selectors.ts

skurfuerst committed:
Merge pull request #3851 from neos/task/use-saucectl

TASK: Use saucectl with configuration

markusguenther committed:
BUGFIX: Use CIRCLE_BRANCH when we are not triggered by a PR (#3854)

markusguenther commented:
As I got the feedback that we have too many shell scripts for the tasks, I would try to unify the scripts somehow.
So that we have on script that can be run saucelabs by a parameter and local if it is not set.

markusguenther committed:
TASK: Refactor script to functions

To be able to reuse parts the change extracts tasks into functions.

markusguenther committed:
TASK: Unify shell scripts for local and saucelabs testing

mhsdesign reviewed (approved):
Thank youuuuuuu so muuuuch :) <3

mhsdesign committed:
TASK: Run tests also on chrome in ci

mhsdesign committed:
BUGFIX: Stabilise test locally to have deterministic inspector groups positioning

When the inspector tab panel is scrolled just enough, so that there\'s enough space, SelectBox contents jump below the SelectBox.

Would fail otherwise because "document" would be before "tests"

mhsdesign committed:
WIP attempt to fix screen size for mac e2e tests

 ✖ Select Boxes - SelectBox opens above in creation dialog if there's not enough space below.

   1) Unable to resize the window because the specified size exceeds the screen size. On macOS, a window cannot be larger than the screen.

      Browser: Chrome 129.0.0.0 / Ventura 13

         21 |        .gt(await ReactSelector('NodeCreationDialog SelectBox').getBoundingClientRectProperty('top'));
         22 |});
         23 |
         24 |test('SelectBox opens above in creation dialog if there\'s not enough space below.', async t => {
         25 |    await t
       > 26 |        .resizeWindow(1200, 768)

mhsdesign committed:
TASK: Cleanup bash script

mhsdesign committed:
TASK: Add exit 1 handling in case of failed test suite

mhsdesign committed:
wip try a valid screen size

> WRN Suite errored. error="job start failed (404): Misconfigured -- Resolution you specified is not available for this OS/browser/version/device combo: OS: 'Mac 13', Browser: 'googlechrome', Version: '129.0.6668.59', Device: 'unspecified', Screen Resolution: '1920x1080'; available resolutions: '1024x768', '1152x864', '1280x960', '1376x1032', '1440x900', '1600x1200', '1920x1440', '2048x1536'"

mhsdesign committed:
TASK: remove retry for now in e2e

The retrys just fail:

Sauce TestCafe Runner 3.9.0
2Running TestCafe 3.6.2

8Error: Cannot find module '/Volumes/Sauce/node/20.14.0/node-v20.14.0-darwin-x64/bin/node_modules/npm/bin/npm-cli.js'

25
26Error: Cannot find module '/Volumes/Sauce/node/20.14.0/node-v20.14.0-darwin-x64/bin/node_modules/npm/bin/npm-cli.js'

56ERROR No tests match your filter.
57See https://testcafe.io/documentation/402638/reference/configuration-file#filter.
58
59Type "testcafe -h" for help.
60JUnit file generation skipped: no test suites detected.

mhsdesign committed:
wip reset screen size

mhsdesign commented:
With testcafe 3 we can use native cdp automation for chromium. So to ensure that our tests also work on chromium based browsers locally we should add chrome to the test matrix. Thats why i change the test to run on chrome on mac os.

Additionally i added exit code handling to the bash script to signal if a test failed.

mhsdesign committed:
wip revert test resize hack `resizeWindow` as mac os on e2e will fail

see 20dfb5a6c50b85e8975733b7b49480d753f9e416
https://github.com/neos/neos-ui/pull/3275

mhsdesign committed:
give up
